azavea/iow-boundary-tool
# Implement login interface

## Overview

Implement login interface.

- Expose endpoints for login, logout, and other standard endpoints from [`dj-rest-auth`](https://github.com/iMerica/dj-rest-auth)
- Add login and logout views at `api/auth/login` and `api/auth/logout`
- Redirect users to login page when logged out
- Redirect users to their originally requested page upon successful login
- Avoid forcing users to login again when refreshing the page  

Closes #66

### Demo
<img width="500" alt="Screen Shot 2022-09-22 at 5 49 25 PM" src="https://user-images.githubusercontent.com/38668450/191857629-f1a6e49d-8165-4321-a7fb-cf2c9a4e29da.png">

```shell
% curl -X POST --data '{"email":"e","password":"p"}' -H "Content-Type:application/json" localhost:8181/api/auth/login/
{"email":["Enter a valid email address."]}

% curl -X POST --data '{"email":"a1@azavea.com","password":"p"}' -H "Content-Type:application/json" localhost:8181/api/auth/login/
{"non_field_errors":["Unable to log in with provided credentials."]}

% curl -X POST --data '{"email":"a1@azavea.com","password":"password"}' -H "Content-Type:application/json" localhost:8181/api/auth/login/
```

### Notes

- So far, I chose not to `await` the `/login` requests, so let me know if that is desired.
- `dj-rest-auth` is the currently maintained fork of `django-rest-auth`, which will not be compatible with Django 4.
- I opted for placeholder texts in the input boxes instead of labels, but I did add `aria-label`s
- I figured it was harmless to retain the `username` field of the /login endpoint: it appears to work just fine (EDIT: well, it works using the dj-rest-auth endpoint, not the session-based django endpoint, and now only the django endpoint is exposed as of 8039f8f4960d031dade9f5859ad3cf671dd2728f). Let me know if it's worth expending effort here to hide the username field.

## Testing Instructions

- `./scripts/update`
- `./scripts/server`
- Visit [localhost:8181/draw](http://localhost:8181/draw)
  - [x] Redirected to login
  - [x] Incorrect or missing credentials are rejected with an error message
  - [x] Successful login reroutes you directly to /draw
- Refresh while logged in
  - [x] You are not rerouted to login screen
- Simulate logout by clearing browser data, refresh page
  - [x] You _are_ rerouted to login screen
  - [x] Successful login reroutes you to /welcome
 - [x] Can log in with Enter key from username or password field
 - [x] Test browsable API for django login and logout views at [localhost:8181/api/auth/login](http://localhost:8181/api/auth/login) and [localhost:8181/api/auth/logout](http://localhost:8181/api/auth/logout)
 - [x] Visit Django admin at [localhost:8181/admin/api](http://localhost:8181/api/admin/api), ensure no additional auth-related entities present
 - [x] Compare styling of login page to figma designs in #66 
 - [x] Ensure login form follows accessibility standards


## Checklist
